### ISSUE TITLE
Controller 1.2.1 failed to start after upgrade

### ISSUE THREAD

**Issue by gen555** [LINBIT/linstor-server#92]
Hi!
After upgrade to Controller 1.2.1 it doesn't start.
Could I downgrade it or how can I fix the issue?

There is log below:

> Dec  1 23:28:43 hyper1 Controller: 23:28:43.822 [Main] INFO  LINSTOR/Controller - SYSTEM - Initializing the database connection pool
> Dec  1 23:28:44 hyper1 Controller: 23:28:44.263 [Main] INFO  org.flywaydb.core.internal.license.VersionPrinter - Flyway Community Edition 5.2.4 by Boxfuse
> Dec  1 23:28:44 hyper1 Controller: 23:28:44.273 [Main] INFO  org.flywaydb.core.internal.database.DatabaseFactory - Database: jdbc:h2:/var/lib/linstor/linstordb (H2 1.4)
> Dec  1 23:28:44 hyper1 Controller: 23:28:44.361 [Main] INFO  org.flywaydb.core.internal.command.DbValidate - Successfully validated 29 migrations (execution time 00:00.034s)
> Dec  1 23:28:44 hyper1 Controller: 23:28:44.383 [Main] INFO  org.flywaydb.core.internal.command.DbMigrate - Current version of schema "LINSTOR": 2019.07.23.13.37
> Dec  1 23:28:44 hyper1 Controller: 23:28:44.383 [Main] WARN  org.flywaydb.core.internal.command.DbMigrate - outOfOrder mode is active. Migration of schema "LINSTOR" may not be reproducible.
> Dec  1 23:28:44 hyper1 Controller: 23:28:44.384 [Main] INFO  org.flywaydb.core.internal.command.DbMigrate - Migrating schema "LINSTOR" to version 2019.09.09.01.01 - Fix resource definition external name entries
> Dec  1 23:28:44 hyper1 Controller: 23:28:44.409 [Main] INFO  org.flywaydb.core.internal.command.DbMigrate - Migrating schema "LINSTOR" to version 2019.10.01.01.01 - Enables auto-quorum and auto-add-tiebreaker on controller level
> Dec  1 23:28:44 hyper1 Controller: 23:28:44.430 [Main] INFO  org.flywaydb.core.internal.command.DbMigrate - Migrating schema "LINSTOR" to version 2019.10.17.01.01 - Delete the remaining properties after a snapshot restore operation
> Dec  1 23:28:44 hyper1 Controller: 23:28:44.444 [Main] INFO  org.flywaydb.core.internal.command.DbMigrate - Migrating schema "LINSTOR" to version 2019.10.31.01.01 - Delete the remaining properties after a snapshot restore operation
> Dec  1 23:28:44 hyper1 Controller: 23:28:44.456 [Main] INFO  org.flywaydb.core.internal.command.DbMigrate - Successfully applied 4 migrations to schema "LINSTOR" (execution time 00:00.087s)
> Dec  1 23:28:44 hyper1 Controller: 23:28:44.460 [Main] INFO  LINSTOR/Controller - SYSTEM - Loading security objects
> Dec  1 23:28:44 hyper1 Controller: 23:28:44.467 [Main] INFO  LINSTOR/Controller - SYSTEM - Current security level is NO_SECURITY
> Dec  1 23:28:44 hyper1 Controller: 23:28:44.479 [Main] INFO  LINSTOR/Controller - SYSTEM - Core objects load from database is in progress
> Dec  1 23:28:44 hyper1 Controller: 23:28:44.623 [Main] ERROR LINSTOR/Controller - SYSTEM - Problem of type 'java.lang.NullPointerException' logged to report number 5DE4227A-00000-000000
> Dec  1 23:28:44 hyper1 Controller: 23:28:44.624 [Thread-1] INFO  LINSTOR/Controller - SYSTEM - Shutdown in progress

> ERROR REPORT 5DE4227A-00000-000000
> 
> ============================================================
> 
> Application:                        LINBIT® LINSTOR
> Module:                             Controller
> Version:                            1.2.1
> Build ID:                           b9f4b7d89f6ec5904f4098c205f36b6fd0dc0078
> Build time:                         2019-11-07T13:10:31+00:00
> Error time:                         2019-12-01 23:28:44
> Node:                               hyper1
> 
> ============================================================
> 
> Reported error:
> ===============
> 
> Category:                           RuntimeException
> Class name:                         NullPointerException
> Class canonical name:               java.lang.NullPointerException
> Generated at:                       Method 'restoreDrbdVolumes', Source file 'DrbdLayerGenericDbDriver.java', Line #536
> 
> 
> Call backtrace:
> 
>     Method                                   Native Class:Line number
>     restoreDrbdVolumes                       N      com.linbit.linstor.core.objects.DrbdLayerGenericDbDriver:536
>     load                                     N      com.linbit.linstor.core.objects.DrbdLayerGenericDbDriver:494
>     loadRscLayerData                         N      com.linbit.linstor.dbdrivers.DatabaseLoader:615
>     loadLayerObects                          N      com.linbit.linstor.dbdrivers.DatabaseLoader:559
>     loadAll                                  N      com.linbit.linstor.dbdrivers.DatabaseLoader:497
>     loadCoreObjects                          N      com.linbit.linstor.core.DbDataInitializer:153
>     initialize                               N      com.linbit.linstor.core.DbDataInitializer:103
>     start                                    N      com.linbit.linstor.core.Controller:242
>     main                                     N      com.linbit.linstor.core.Controller:597
> 
> 
> END OF ERROR REPORT.
>

**Comment by ghernadi** [comment]
You can try downgrading your system, yes, but if you do that, do not forget to also restore the old database. It looks like you are using the default settings for the database, which means the backed-up database should be in the same directory as the current database (`/var/lib/linstor/`) - look for the correct time-suffix to rollback to the correct version.

Additionally I would appreciate if you could send me both, the backup as well as the currently broken database file

**Comment by gen555** [comment]
[https://dropmefiles.com/NSzE4](all DBs)

**Comment by ghernadi** [comment]
Thank you.

If you have upgraded from <= `v1.0.1`, then we already have fixed the actual error (namely that we did not remove some entries in the database when removing a resource, restarting the controller afterwards fails because those entries point to objects that got already deleted...), but apparently we forgot writing a migration script to cleanup already broken databases...

I am not sure if downgrading actually lets you start the controller or if your old version also crashes with a similar error.

We will write the migration to prevent this error report in the future. Until then you can execute this SQL on your database:   
`DELETE FROM LINSTOR.LAYER_DRBD_VOLUMES WHERE LAYER_RESOURCE_ID = 52;`  
You can execute the SQL with something like
```
echo 'DELETE FROM LINSTOR.LAYER_DRBD_VOLUMES WHERE LAYER_RESOURCE_ID = 52;' | /usr/share/linstor-server/bin/linstor-config sql-script /etc/linstor/linstor.toml
```
(adapt the paths if you are not using default installation paths)

**Comment by gen555** [comment]
Thanks!
I'll give it a try.

**Comment by ghernadi** [comment]
Great - keep me updated if it works or not

ps.: if you want you can delete the link to the database if you do not want to have it publicly visible

**Comment by gen555** [comment]
It worked. :)

**Comment by ghernadi** [comment]
Glad it worked. For the others: the in the meantime released version `v1.3.0` and higher already contain a migration that should fix all issues of this kind. No more need for the above mentioned manual SQL workaround :)
